### PR TITLE
mcs: fix ticker stop in register

### LINKED ISSUE #6875: Keep alive ticker is wrongly stopped
## Bug Report

related to https://github.com/tidbcloud/pd-cse/issues/153.

https://github.com/tikv/pd/blob/0c537bb5bfcea365248f290bc126cd584e5ef337/pkg/mcs/discovery/register.go#L84-L107

Line 98 and line 104 are wrongly stopped by #6720.

### What version of PD are you using (`pd-server -V`)?

cse only

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #6875

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
pd0: `./pd-server services api`
tso1: `./pd-server services tso --listen-addr="http://127.0.0.1:3380" --backend-endpoints="http://127.0.0.1:2379"`
tso2: `./pd-server services tso --listen-addr="http://127.0.0.1:3379" --backend-endpoints="http://127.0.0.1:2379"`

can find register key
image

then close pd0 and wait a minutes and can see  keepalive failed log
image

and restart pd0 and can see normal and no keepalive failed log anymore
image
image

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/mcs/discovery/register.go [hunk 1]
```diff
@@ -82,11 +82,11 @@ func (sr *ServiceRegister) Register() error {
 					log.Error("keep alive failed", zap.String("key", sr.key))
 					// retry
 					t := time.NewTicker(time.Duration(sr.ttl) * time.Second / 2)
+					defer t.Stop()
```
Comment (rleungx) [inline pkg/mcs/discovery/register.go:85]: We cannot use defer here.

File: pkg/mcs/discovery/register.go [hunk 2]
```diff
@@ -73,37 +73,43 @@ func (sr *ServiceRegister) Register() error {
...
+				if ok {
+					continue
+				}
+				log.Error("keep alive failed", zap.String("key", sr.key))
+				// retry
+				t := time.NewTicker(time.Duration(sr.ttl) * time.Second / 2)
+				for {
+					select {
+					case <-sr.ctx.Done():
+						log.Info("exit register process", zap.String("key", sr.key))
```
Comment (JmPotato) [inline pkg/mcs/discovery/register.go:91]: What about distinguishing from line 79?

File: pkg/mcs/discovery/register.go [hunk 3]
```diff
@@ -73,37 +73,43 @@ func (sr *ServiceRegister) Register() error {
...
-							t.Stop()
 							continue
 						}
+
+						kresp, err = sr.cli.KeepAlive(sr.ctx, resp.ID)
+						if err != nil {
+							log.Error("client keep alive failed", zap.String("key", sr.key), zap.Error(err))
+							continue
+						}
+						t.Stop()
```
Comment (JmPotato) [inline pkg/mcs/discovery/register.go:111]: What about extracting a function returns `kresp`? It's kind of annoying to introduce a nested `for-select` block here.

**Review by JmPotato — APPROVED** [review]

File: pkg/mcs/discovery/register_test.go [hunk 4]
```diff
@@ -49,20 +45,40 @@ func TestRegister(t *testing.T) {
...
 	re.NoError(err)
-	re.Empty(resp.Kvs)
+	re.Equal("127.0.0.1:2", getKeyAfterLeaseExpired(re, client, sr.key))
+	etcd.Close()                // close the etcd to make the keepalive failed
+	time.Sleep(3 * time.Second) // ensure that the lease is expired
+	etcd, err = embed.StartEtcd(cfg)
+	re.NoError(err)
+	<-etcd.Server.ReadyNotify()
+	defer etcd.Close()
+	re.Equal("127.0.0.1:2", getKeyAfterLeaseExpired(re, client, sr.key))
```
Comment (rleungx) [inline pkg/mcs/discovery/register_test.go:73]: The test can also pass in master.

**Comment by lhy1024** [comment]
> The test can also pass in the master branch.

Thanks, I have ensured master branch cannot pass with new test

image

**Review by rleungx — APPROVED** [review]

File: pkg/mcs/discovery/register.go [hunk 5]
```diff
@@ -113,6 +86,32 @@ func (sr *ServiceRegister) Register() error {
...
+	t := time.NewTicker(time.Duration(sr.ttl) * time.Second / 2)
+	defer t.Stop()
+	for {
+		select {
+		case <-sr.ctx.Done():
+			log.Info("exit register process", zap.String("key", sr.key))
+			return nil
+		case <-t.C:
+			ctx, cancel := context.WithTimeout(sr.ctx, etcdutil.DefaultRequestTimeout)
+			defer cancel()
```
Comment (rleungx) [inline pkg/mcs/discovery/register.go:99]: defer cancel?

**Pull request merged**